zeus/optimizer/batch_size/client.py

```diff
@@ -0,0 +1,115 @@
+from __future__ import annotations
+
+from abc import ABC, abstractmethod
+from uuid import UUID
+
+import pynvml
+import httpx
+
+from zeus.callback import Callback
+from zeus.job import Job
+from zeus.monitor import ZeusMonitor
+from zeus.monitor.energy import Measurement
+from zeus.optimizer.batch_size.server.models import (
+    JobSpec,
+    ReportResponse,
+    TrainingResult,
+)
+from zeus.util.logging import get_logger
+from zeus.util.metric import zeus_cost
+
+"""
+TODO: Do we want one job -> one BSO Client? 
+"""
+
+
+class BatchSizeOptimizerClient(Callback):
+    def __init__(self, monitor: ZeusMonitor, server_url: str, job: JobSpec) -> None:
+
+        self.monitor = monitor
+        self.server_url = server_url
+        self.job = job
+        self.cur_epoch = 0
+        self.running_time = 0.0
+        self.consumed_energy = 0.0
+
+        # Get max PL
+        pynvml.nvmlInit()
+        pls = []
+        self.max_power = 0
+        for index in self.monitor.nvml_gpu_indices:
+            device = pynvml.nvmlDeviceGetHandleByIndex(index)
+            pls.append(pynvml.nvmlDeviceGetPowerManagementLimitConstraints(device))
+        if not all(pls[0] == pl for pl in pls):
+            raise ValueError("Power limits ranges are not uniform across GPUs.")
+
+        self.max_power = max(pls) * len(monitor.gpu_indices)
+        self.current_batch_size = 0
+
+        # Register job
+        res = httpx.post(self.server_url + "/jobs", content=job.json())
+        self._handle_response(res)
+
+    def get_batch_size(self) -> int:
+        """Get batch size to use from the BSO server"""
+        self.cur_epoch = 0
+        res = httpx.get(
+            self.server_url + "/jobs/batch_size", params={"job_id": self.job.job_id}
```

Comment:
For constants like `/jobs/batch_size`, you should create a `zeus.optimizer.batch_size.common` that holds all these. Then, both the client and server should import from it. For instance, `GET_BATCH_SIZE_PATH = "/jobs/batch_size"`.